### PR TITLE
333 - Add support for native indexers

### PR DESCRIPTION
**Pull request by binarycow**

### CODE REVIEW

File: src/Scriban/Syntax/Expressions/ScriptIndexerExpression.cs [hunk 1]
```diff
@@ -130,17 +130,22 @@ private object GetOrSetValue(TemplateContext context, object valueToSet, bool se
             {
                 var accessor = context.GetMemberAccessor(targetObject);
                 var indexAsString = context.ObjectToString(index);
+                var itemAccessor = accessor as IItemAccessor;
 
                 if (setter)
                 {
-                    if (!accessor.TrySetValue(context, Index.Span, targetObject, indexAsString, valueToSet))
+                    var itemSuccess = itemAccessor?.ItemType == index?.GetType() && itemAccessor.TrySetItem(context, Index.Span, targetObject, index, valueToSet);
```
Comment (xoofx) [inline src/Scriban/Syntax/Expressions/ScriptIndexerExpression.cs:137]: Should probably use `context.ToObject` to `itemAccessor?.ItemType` instead to convert the index only if ItemType is not null and use the item accessor only in this case.

File: src/Scriban/Runtime/IItemAccessor.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// Licensed under the BSD-Clause 2 license.
+// See license.txt file in the project root for full license information.
+#nullable disable
+
+using System;
+using Scriban.Parsing;
+
+namespace Scriban.Runtime
+{
+    public interface IItemAccessor
```
Comment (xoofx) [inline src/Scriban/Runtime/IItemAccessor.cs:11]: Maybe add a comment to explain that it is for accessing indexers on object.

File: src/Scriban/Syntax/Expressions/ScriptIndexerExpression.cs (hunk 1 above)
Comment (xoofx) [reply src/Scriban/Syntax/Expressions/ScriptIndexerExpression.cs:137]: (a source indexer value could be e.g long while the indexer is int, automatic casting is usually done in scriban for such case)

Comment (binarycow) [reply src/Scriban/Syntax/Expressions/ScriptIndexerExpression.cs]: Updated

File: src/Scriban/Runtime/IItemAccessor.cs (hunk 2 above)
Comment (binarycow) [reply src/Scriban/Runtime/IItemAccessor.cs:15]: Updated

**Comment by xoofx** [comment]
LGTM, could you add a test for this support? Thanks!

**Comment by xoofx** [comment]
It seems that some existing tests are failing.

**Comment by binarycow** [comment]
Yep, workin on 'em.  Gotta prioritize my day job and all!

**Comment by binarycow** [comment]
@xoofx - Try it now; tests are passing on my local machine.

**Comment by xoofx** [comment]
Thanks!

**Pull request merged**